tests/suites/test_suite_ssl.function

```diff
@@ -361,6 +674,291 @@ static int ssl_populate_session( mbedtls_ssl_session *session,
  * END_DEPENDENCIES
  */
 
+/* BEGIN_CASE */
+void test_callback_buffer( int size, int put1, int put1_ret,
+                           int get1, int get1_ret, int put2, int put2_ret,
+                           int get2, int get2_ret )
+{
+#define ROUNDS 2
+    size_t put[ROUNDS];
+    int put_ret[ROUNDS];
+    size_t get[ROUNDS];
+    int get_ret[ROUNDS];
+    mbedtls_test_buffer buf;
+    unsigned char* input = NULL;
+    size_t input_len;
+    unsigned char* output = NULL;
+    size_t output_len;
+    size_t i, written, read;
+    int j;
+
+    mbedtls_test_buffer_init( &buf );
+    TEST_ASSERT( mbedtls_test_buffer_setup( &buf, size ) == 0 );
+
+    TEST_ASSERT( put1 >= 0 );
+    put[0] = put1;
+    put_ret[0] = put1_ret;
+    TEST_ASSERT( put1_ret <= put1 );
+    TEST_ASSERT( put2 >= 0 );
+    put[1] = put2;
+    put_ret[1] = put2_ret;
+    TEST_ASSERT( put2_ret <= put2 );
+
+    TEST_ASSERT( get1 >= 0 );
+    get[0] = get1;
+    get_ret[0] = get1_ret;
+    TEST_ASSERT( get1_ret <= get1 );
+    TEST_ASSERT( get2 >= 0 );
+    get[1] = get2;
+    get_ret[1] = get2_ret;
+    TEST_ASSERT( get2_ret <= get2 );
+
+    /* Calculate actual input and output lengths */
```

Comment:
Why not make `put` and `get` buffers (our test suite's `data_t` type) and get the length from there?

Reply:
I am sorry I am not sure I understand how that would work. Could you please elaborate?

Reply:
My initial reading of this code was that it looks to see that you are passing in 0 for put2 and get2 to deterimine the number of rounds. However, I was mistaken. You are summing up the values, not counting the number of values.

That said, we could simplify this test code quite a bit by determining the number of rounds from `data_t` inputs. You wouldn't need local `put` or `get` buffers on the stack either, if you can use `data_t`.

```C
void test_callback_buffer( int size, data_t *put, data_t *put_ret,
                           data_t *get, data_t *get_ret )
```

Then, you could do `put->len` to get the number of rounds (and it'd probably be wise to assert all inputs are the same length), and can easily scale up to more than 2 rounds if desired without tons of function parameters. You can skip copying input parameters to an array, and can do `put->x[0]` to read the data.  A lot of code from this test could be removed if you were to use `data_t`, and that would help readability.

Reply:
Not blocking: just a suggestion

Reply:
I considered using `data_t` parameters, but I thought that with `int` parameters it is easier to write and read the test cases.

I can see your point, using `data_t` would make the test function significantly shorter. However, the code we could omit is simple and this limits the impact on readability. Therefore I still would prefer to go here for test cases that are easier readable/writable.